### ISSUE TITLE
Server runs out of memory when working with WireData indexes

### ISSUE THREAD

**Issue by zoaked** [processwire/processwire-issues#485]
### Short description of the issue
Unable to use numerical indexes on WireData objects

More detailed description of the issue:
PHP will helpfully convert "0" into 0 when used inside array notation, but the same problem will exist if you set an integer key to any value.  In PHP, all strings are converted to 0 when compared with an integer.  (`$left = 0; $right = 'data'; var_dump($left == $right); // bool(true)`)  This means that when WireData tries to save the assignment it results in a false positive match and tries to replace the entire object in some kind of recursive fashion.

### Expected behavior
Numerical indexes behave the same as native arrays

### Actual behavior
Server runs out of memory

### Optional: Suggestion for a possible fix
\- PARTIAL FIX! - 
WireData->set() performs a loosely typed check (==) to see if the key is data.  Need to use a strong typed comparison (===) so the values do not get converted from their native types.
https://github.com/processwire/processwire/blob/master/wire/core/WireData.php#L62

NOTE: There are several more places where $var == 'data' or $var == "data" is used and they would all likely have the same issue.

The above item solves our immediate need by avoiding the real bug.  I'm not sure what this logic is supposed to do or when it would be used, but there is still the underlying problem of getting stuck in some kind of loop when setArray is called. (set method is called over 4k more times after that which is what causes the server to run out of memory) 

### Steps to reproduce the issue
1. Add the following code to any template file (I used basic-page) that is used on at least one page
`		<?php
			$obj = new \processwire\WireData();
			$obj[0] = "Text";
			/* or */
			$obj->set(0, "Text");
		?>
`
2. Visit a page that uses this template
3. Watch the world burn (server runs out of memory)

### Setup/Environment

* ProcessWire: 3.0.62
* PHP: 7.0.22-0ubuntu0.16.04.1
* Webserver: Apache/2.4.18 (Ubuntu)
* MySQL: 5.7.17

**Comment by ryancramerdesign** [comment]
Thanks @zoaked. WireData is associative only, it isn't intended to be used with numerical indexes (WireArray is what you want instead), but of course we don't want attempting to use a numeric index causing the error you've highlighted, so it definitely needs to be fixed.  Thanks for your proposed solution, it seems to resolve the issue here. Though you mentioned it was a "partial fix", what do you mean by that?

**Comment by zoaked** [comment]
I had attempted to convey why using strict type comparison was listed as a partial fix in the 3rd paragraph of the proposed solution section.

Something made the logic within that if data block run recursively out of control. The concern is that the looping issue was separate and might still get triggered when 'data' was passed as a key. When this issue was identified, we stopped looking so do not have an example.
